### PR TITLE
feat: range request support

### PR DESCRIPTION
**Pull request by Arqu**
Closes https://github.com/n0-computer/iroh/issues/130

### CODE REVIEW

File: iroh-gateway/src/headers.rs [hunk 1]
```diff
@@ -74,12 +74,41 @@ pub fn add_content_disposition_headers(
...
     headers.insert(
         CONTENT_DISPOSITION,
         HeaderValue::from_str(&format!("{}; filename={}", disposition, filename)).unwrap(),
     );
 }
 
+#[tracing::instrument()]
+pub fn add_content_range_headers(headers: &mut HeaderMap, range: Range<u64>, size: Option<u64>) {
+    let content_range = if let Some(size) = size {
+        format!("bytes {}-{}/{}", range.start, range.end - 1, size)
```
Comment (dignifiedquire) [inline iroh-gateway/src/headers.rs:87]: if `range.end`  is `0` this will panic

File: iroh-gateway/src/headers.rs [hunk 2]
```diff
@@ -74,12 +74,41 @@ pub fn add_content_disposition_headers(
...
+    } else {
+        format!("bytes {}-{}/{}", range.start, range.end - 1, "*")
+    };
+    headers.insert(
+        CONTENT_RANGE,
+        HeaderValue::from_str(&content_range).unwrap(),
+    );
+}
+
+pub fn parse_range_header(range: &HeaderValue) -> Option<Range<u64>> {
```
Comment (dignifiedquire) [inline iroh-gateway/src/headers.rs:97]: should have tests for parsing and generating these headers

File: iroh-resolver/src/resolver.rs [hunk 3]
```diff
@@ -85,6 +85,22 @@ impl Path {
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+pub enum ResponseClip {
```
Comment (dignifiedquire) [inline iroh-resolver/src/resolver.rs:89]: docs please

File: iroh-resolver/src/resolver.rs [hunk 4]
```diff
@@ -480,6 +516,41 @@ impl<T: ContentLoader + Unpin + 'static> AsyncRead for OutPrettyReader<T> {
...
+            | OutPrettyReader::DagJson(bytes_reader)
+            | OutPrettyReader::Raw(bytes_reader) => {
+                let pos_current = bytes_reader.pos;
+                let data_len = bytes_reader.bytes.len();
+                match position {
+                    std::io::SeekFrom::Start(pos) => {
+                        let i = std::cmp::min(data_len, pos as usize);
+                        bytes_reader.pos = i;
+                    }
+                    std::io::SeekFrom::End(_pos) => {
```
Comment (dignifiedquire) [inline iroh-resolver/src/resolver.rs:533]: I believe `End` also supports negative seek, eg `-5` which should be easily supported in this case

File: iroh-resolver/src/resolver.rs [hunk 5]
```diff
@@ -480,6 +516,41 @@ impl<T: ContentLoader + Unpin + 'static> AsyncRead for OutPrettyReader<T> {
...
+                        // This technically means we can seek past the end which we don't support
+                        // TODO: support rolling buffer reads, ie use overflow remainder to read from the start again
+                        todo!()
+                    }
+                    std::io::SeekFrom::Current(pos) => {
```
Comment (dignifiedquire) [inline iroh-resolver/src/resolver.rs:538]: `pos` can be negative, just casting it to `usize` is not enough

File: iroh-resolver/src/resolver.rs [hunk 6]
```diff
@@ -1417,9 +1496,13 @@ mod tests {
             if let OutContent::Unixfs(node) = ipld_hello_txt.content {
                 assert_eq!(
                     read_to_string(
```
Comment (dignifiedquire) [inline iroh-resolver/src/resolver.rs:1498]: Should have tests with clipping and seeking

File: iroh-resolver/src/unixfs.rs [hunk 7]
```diff
@@ -492,16 +498,57 @@ impl<T: ContentLoader + Unpin + 'static> AsyncRead for UnixfsContentReader<T> {
...
+                        *pos = (*pos as i64 + offset) as usize;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn poll_complete(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<std::io::Result<u64>> {
+        Poll::Ready(Ok(0))
```
Comment (dignifiedquire) [inline iroh-resolver/src/unixfs.rs:532]: wrong return value

File: iroh-gateway/src/headers.rs (hunk 1 above)
Comment (Arqu) [reply iroh-gateway/src/headers.rs:87]: I can add another guard, but this path is not invoked unless range exists and is passed in. And that is checked against in `parse_range_header`

Comment (dignifiedquire) [reply iroh-gateway/src/headers.rs:87]: then maybe add a comment about that

File: iroh-resolver/src/resolver.rs [hunk 8]
```diff
@@ -480,6 +516,41 @@ impl<T: ContentLoader + Unpin + 'static> AsyncRead for OutPrettyReader<T> {
...
+                    }
+                }
+                Ok(())
+            }
+            OutPrettyReader::Unixfs(r) => Pin::new(&mut *r).start_seek(position),
+        }
+    }
+
+    fn poll_complete(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<std::io::Result<u64>> {
+        Poll::Ready(Ok(0))
```
[the remark this answers is not in the document]
Comment (Arqu) [reply iroh-resolver/src/resolver.rs:550]: Tbh I don't fully understand this part around poll_complete and I just pieced it together from tokio's BufReader. 
The result if not `(0)` is `self.as_mut().get_pin_mut().poll_complete(cx)?` which to me in their case seems to resolve to the same.

File: iroh-resolver/src/unixfs.rs (hunk 7 above)
Comment (Arqu) [reply iroh-resolver/src/unixfs.rs:532]: Same as above, would love some input on what it should be. I'm assuming the delta from the initial position which in a simplified version returns the seek offset?

File: iroh-resolver/src/resolver.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply iroh-resolver/src/resolver.rs:550]: Even it does resolve to the same, that's what it relies on for the seeking, it should proxy there.

File: iroh-resolver/src/unixfs.rs (hunk 7 above)
Comment (dignifiedquire) [reply iroh-resolver/src/unixfs.rs:532]: Yes, I think this is the delta, so we need to store it and return it. Later when we do seeking with loading, this will also delegate to an internal future.

Comment (Arqu) [reply iroh-resolver/src/unixfs.rs:537]: So I took some time to understand this and as far as I get it it's not the delta but rather the final position it seeked to. Because otherwise the consumer would need to figure out what the delta means especially in circular buffer and negative moves etc.

**Comment by Arqu** [comment]
@dignifiedquire can you take another look at the unixfs seeking, it's funkier now. 
Also added tests and cleaned it up.

File: iroh-resolver/src/resolver.rs [hunk 9]
```diff
@@ -1678,6 +1700,129 @@ mod tests {
...
+
+            // seek and clip
+            let cr = seek_and_clip(&node, resolver.clone(), 1..3).await;
+            assert_eq!(read_to_string(cr).await, "el");
+        } else {
+            panic!("invalid result: {:?}", ipld_hello_txt);
+        }
+    }
+
+    #[tokio::test]
```
Comment (dignifiedquire) [inline iroh-resolver/src/resolver.rs:1759]: Either in this PR or at least have an issue for seeking across dag-cbor, dag-json and raw files as well please.

File: iroh-resolver/src/unixfs.rs (hunk 7 above)
Comment (dignifiedquire) [reply iroh-resolver/src/unixfs.rs]: Yes new offset from the start is what should be returned 👍

File: iroh-resolver/src/resolver.rs (hunk 9 above)
Comment (Arqu) [reply iroh-resolver/src/resolver.rs:1759]: https://github.com/n0-computer/iroh/issues/346

Comment (Arqu) [reply iroh-resolver/src/resolver.rs:1759]: Gotta do other stuff before release :/

**Comment by dignifiedquire** [comment]
merge conflict 😅

**Review by dignifiedquire — DISMISSED** [review]
Looks good, please also open an issue to add some additional tests on the seeking math, that is easy to get wrong on edge cases and I don't trust my own review of them. Some nice proptests should help for confidence.
